File: aoc2022/day04/day04.py
def is_fully_contained(r1, r2):
    return (r1[0] <= r2[0] and r1[1] >= r2[1]) or (r2[0]<= r1[0] and r2[1]>= r1[1])

def are_overlapping(r1, r2):
    return (r1[1] >= r2[0] and r1[0] <= r2[1] ) or (r2[1] >= r1[0] and r2[0] <= r1[1])

data = []
# with open('sample.txt','r') as data_file:
with open('input.txt','r') as data_file:
    for line in data_file.readlines():
        data.append(list([int(x) for x in couple.split('-')] for couple in line.strip().split(',')))

fully_contained_count = 0
partial_overlap_count = 0

for pairs in data:
    if is_fully_contained(pairs[0], pairs[1]):
        fully_contained_count += 1

    if are_overlapping(pairs[0], pairs[1]):
        partial_overlap_count += 1

    # print(pairs, is_fully_contained(pairs[0], pairs[1]), are_overlapping(pairs[0], pairs[1]))

print(f'Part 1: {fully_contained_count}') # 573

print(f'Part 2: {partial_overlap_count}') # 867

File: aoc2022/day04/day04.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

func Solve() (int, int) {
	dataFile, err := os.Open("input.txt")
	if err != nil {
		log.Fatal("failed to open data file")
	}

	fullyOverlapCount := 0
	partialOverlapCount := 0

	type IdRange struct {
		start int
		end   int
	}

	var elf1 IdRange
	var elf2 IdRange

	fileScanner := bufio.NewScanner(dataFile)
	for fileScanner.Scan() {
		fmt.Sscanf(fileScanner.Text(), "%d-%d,%d-%d", &elf1.start, &elf1.end, &elf2.start, &elf2.end)

		if (elf1.start <= elf2.start && elf1.end >= elf2.end) || (elf2.start <= elf1.start && elf2.end >= elf1.end) {
			fullyOverlapCount++
		}

		if (elf1.end >= elf2.start && elf1.start <= elf2.end) || (elf2.end >= elf1.start && elf2.start <= elf1.end) {
			partialOverlapCount++
		}
	}

	dataFile.Close()

	return fullyOverlapCount, partialOverlapCount
}
